File: lru-cache/lru-cache.java
class ListNode {
    int key;
    int val;
    ListNode prev;
    ListNode next;
    public ListNode(int key, int val) {
        this.key = key;
        this.val = val;
    }
}

class LRUCache {
    Map<Integer, ListNode> map;
    int capacity;
    ListNode sHead;
    ListNode sTail;
    public LRUCache(int capacity) {
        map = new HashMap<>();
        this.capacity = capacity;
        sHead = new ListNode(-1, -1);
        sTail = new ListNode(-1, -1);
        sHead.next = sTail;
        sTail.prev = sHead;
    }
    
    public int get(int key) {
        if (!map.containsKey(key)) return -1;
        ListNode node = map.get(key);
        int result = node.val;
        remove(node);
        add(node);
        return result;
    }
    
    public void put(int key, int value) {
        if (map.containsKey(key)) {
            ListNode node = map.get(key);
            remove(node);
        }
        ListNode newNode = new ListNode(key, value);
        add(newNode);
        map.put(newNode.key, newNode);

        if (map.size() > capacity) {
            ListNode nodeToEvict = sHead.next;
            remove(nodeToEvict);
            map.remove(nodeToEvict.key);
        }
    }

    private void remove(ListNode node) {
        //map.remove(node.key);
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void add(ListNode node) {
        //map.put(node.key, node);
        sTail.prev.next = node;
        node.prev = sTail.prev;
        node.next = sTail;
        sTail.prev = node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146-lru-cache/lru-cache.py
class Node:
    def __init__(self, key, val: int):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None
class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {}
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self.remove(node)
        self.add(node)
        return node.val
        

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            self.remove(node)
        nodeToAdd = Node(key, value)
        self.add(nodeToAdd)
        self.cache[key] = nodeToAdd
        if len(self.cache) > self.capacity:
            nodeToRemove = self.head.next
            self.remove(nodeToRemove)
            del self.cache[nodeToRemove.key]
    def add(self, node):
        lastNode = self.tail.prev
        lastNode.next = node
        node.prev = lastNode
        node.next = self.tail
        self.tail.prev = node
    
    def remove(self, node):
        prevNode = node.prev
        nextNode = node.next
        prevNode.next = nextNode
        nextNode.prev = prevNode


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)